### ISSUE TITLE
bug: Environment variables from LOVELY_HELM_MERGE are expanded

### ISSUE THREAD

**Issue by dennislapchenko** [crumbhole/argocd-lovely-plugin#217]
I would assume that environment variables that can be part of my values configuration should not expanded.
I have a manifest like so:
```yaml
spec:
    ...
    plugin:
      name: argocd-lovely-plugin
      env:
        - name: LOVELY_HELM_MERGE
          value: |
            serviceAccount:
              annotations:
                test: yes
                simplevar: $da
                quoted_var: "$AVP_TYPE"
                escaped_var: "\\$AVP_TYPE"
    ...
```

The resulting resource looks like this:
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    escaped_var: \
    quoted_var: ""
    test: "yes"
```

Escaping a value once, like `"\$AVP_TYPE"` breaks yaml altogether and i get this:
```
Unknown desc = error generating manifests: `argocd-lovely-plugin` failed exit status 1: 2023/07/03 15:18:41 yaml: line 2: did not find expected key"
```

I am not sure this is intended behaviour as there is no declared env var expansion functionality.
This totally breaks all application configurations that have env vars that the apps process by their logic (primarily its the dex configuration for argocd.. it uses `clientID: $dex.gitlab.clientID` this format for config and its broken when using lovely)

Expected behaviour:
- Ignore all environment variables?

**Comment by Joibel** [comment]
https://github.com/argoproj/argo-cd/issues/12152 Says that multiline env vars are broken, yet yours seems to be working better than expected. Have you tried doing this with a single line first to see if anything works that way?

Lovely itself isn't doing any expansion, it never treats variables in a way that can expand that I can think of. What is the expected source of AVP_TYPE (environment variable on the container?), and how are you running lovely (sidecar or configmap)

**Comment by Joibel** [comment]
By "Lovely not doing expansion" I am suggesting that maybe ArgoCD is doing the expansion before it arrives. I can look into this, but not today.

**Comment by dennislapchenko** [comment]
@Joibel I appreciate your fast reply!

I am not expecting these env vars to be expanded at all. `AVP_TYPE` is just an env that is present in the sidecar, as I am planning to configure argocd-vault-plugin further down the line. (for now that plugin is disabled)

Running lovely via sidecar, standard config:
```yaml
    - name: lovely-tmp
      emptyDir: {}
  extraContainers:
    - name: lovely-plugin
      command: [/var/run/argocd/argocd-cmp-server]
      image: ghcr.io/crumbhole/argocd-lovely-plugin-cmp:0.19.6
      env:
          - name: AVP_TYPE
            value: kubernetessecret
      securityContext:
        runAsNonRoot: true
        runAsUser: 999
      volumeMounts:
        - mountPath: /var/run/argocd
          name: var-files
        - mountPath: /home/argocd/cmp-server/plugins
          name: plugins
        - mountPath: /tmp
          name: lovely-tmp
 ```

When NOT using lovely, the variables arent touched at all, because I presume they are parsed differently, because they come in via
```yaml
spec:
  helm:
    values: |
      ...
```

**Comment by dennislapchenko** [comment]
Plain argocd on its own indeed does not expand variables:

manifest:
```yaml
spec:
  source:
    ...
    helm:
      values: |
        serviceAccount:
          annotations:
            tast: "yes"
            simplevar: '$da'
            quoted_var: "$AVP_TYPE"
            quoted_var2: $dex.huex
            quoted_var5: '\\$dex.huex'
            eshe_odin_ok: "ok$prosto"
```

Result:
```yaml
kind: ServiceAccount
metadata:
  annotations:
    escaped_var: \$AVP_TYPE
    eshe_odin_ok: ok$prosto
    quoted_var2: $dex.huex
    quoted_var5: \\$dex.huex
    simplevar: $da
    tast: "yes"
```

**Comment by Joibel** [comment]
I never expected ArgoCD directly to have problems - it's somewhere in the transport between your application.yaml and the plugin running.

Anyway if you double $ your desired dollars as `$$` they get to where you want as a single `$`
```
spec:
    ...
    plugin:
      name: argocd-lovely-plugin
      env:
        - name: LOVELY_HELM_MERGE
          value: |
            serviceAccount:
              annotations:
                simpledbl: $$da
                quoted_dbl: "$$AVP_TYPE"
                escaped_dbl: "\\$$AVP_TYPE"
    ...
```

and you get 

```
    escaped_dbl: \$AVP_TYPE                                                                                                                                                                                                                    
    quoted_dbl: $AVP_TYPE                                                                                                                                                                                                                      
    simpledbl: $da   
```

**Comment by dennislapchenko** [comment]
@Joibel yeah cheers, forgot that can escape with another `$`. This helps.
However regarding the expansion as a whole - even if its in transport between rendered manifests and the plugin - its still outside of users (my) control, especially since no function like `find` or `generate` needs to be provided when using lovely.

**Comment by Joibel** [comment]
I am confused now as to the problem you're facing.

Lovely is configured using environment variables, and those environment variables need `$$` in order to produce `$`.

Lovely's entire purpose in life is to render manifests, it does that, and nothing else. That is all plugin's in Argo CD can do. Those rendered manfiests are not affected by this `$` problem as far as I can see, unless you are telling me otherwise (please raise a separate bug with examples).

I don't really know how `find` or `generate` come into the picture at all. Generate is a way of configuring plugins, but you can conifgure lovely to do different kinds of work with the preprocessing and plugin parameters. These can be 'optional' work, by having the executable/script they call make decisions based upon any criteria you like, including running `find`.

Can you describe what you're trying to do so I can understand it?

**Comment by dennislapchenko** [comment]
@Joibel My thoughts regarding `find` or `generate` were an attempt to understand what you mean by `it's somewhere in the transport`. If that is now what you meant, ignore.

What i was trying to do is install an Application what contained `$vars` in its inline values configuration.
I installed it via plain argocd and it went all good.
I installed it after enabling lovely and `$vars` became empty strings, from which i deduced that lovely must be expanding variables, in places where i did not want it to be so. I have pasted by tests depicting this here.

TLDR; I wanted to keep manifests as they were, but after enabling lovely, env vars were needlessly expanded inside inline helm configurations (through `LOVELY_HELM_MERGE`).

**Comment by Joibel** [comment]
ArgoCD is doing this expansion, it's not a bug in lovely, rather a feature of ArgoCD
